zerver/lib/users.py

```diff
@@ -501,6 +510,120 @@ def format_user_row(
     return result
 
 
+def check_user_can_access_all_users(acting_user: Optional[UserProfile]) -> bool:
+    if acting_user is None:
+        # We allow spectators to access all users since they
+        # have very limited access to the user already.
+        return True
+
+    if not acting_user.is_guest:
+        return True
+
+    realm = acting_user.realm
+    if is_user_in_group(realm.can_access_all_users_group, acting_user):
+        return True
+
+    return False
+
+
+def get_subscribers_of_target_user_subscriptions(
+    target_users: List[UserProfile],
+) -> Dict[int, Set[int]]:
+    target_user_ids = [user.id for user in target_users]
+    target_user_subscriptions = (
+        Subscription.objects.filter(
+            user_profile__in=target_user_ids,
+            active=True,
+            recipient__type__in=[Recipient.STREAM, Recipient.HUDDLE],
+        )
+        .order_by("user_profile_id")
+        .values("user_profile_id", "recipient_id")
+    )
+
+    target_users_subbed_recipient_ids = set()
+    target_user_subscriptions_dict: Dict[int, Set[int]] = defaultdict(set)
+
+    for user_profile_id, sub_rows in itertools.groupby(
+        target_user_subscriptions, itemgetter("user_profile_id")
+    ):
+        recipient_ids = {row["recipient_id"] for row in sub_rows}
+        target_user_subscriptions_dict[user_profile_id] = recipient_ids
+        target_users_subbed_recipient_ids |= recipient_ids
+
+    subs_in_target_user_subscriptions_query = Subscription.objects.filter(
+        recipient_id__in=list(target_users_subbed_recipient_ids),
+        active=True,
+    )
+
+    subs_in_target_user_subscriptions_query = subs_in_target_user_subscriptions_query.filter(
+        Q(recipient__type=Recipient.STREAM, is_user_active=True)
+        | Q(recipient__type=Recipient.HUDDLE)
+    )
+
+    subs_in_target_user_subscriptions = subs_in_target_user_subscriptions_query.order_by(
+        "recipient_id"
+    ).values("user_profile_id", "recipient_id")
```

Comment:
We will probably want a specialized version of this query for the single-target-user case, since that's what we use in the high throughput code paths like presence. The key optimization is that that we can do something like `.values_list("user_profile_id", flat=True).distinct()` -- i.e. not sending multiple copies from the database to Python for a single `recipient_id`

Reply:
But I think this can be a follow-up optimization; going to merge through this commit so we can get the migrations integrated.

Reply:
@sahil839 is this optimization implemented? Wondering if we need a follow-up issue.